src/graph.rs

```diff
@@ -1655,18 +1731,25 @@ pub(crate) struct AssertTypeWithRange {
   kind: String,
 }
 
+/// Metadata for modules which should suffice to check assertions on importers
+/// of that module. Needs to be stored separately from `Builder::graph` for
+/// concurrent borrows.
+#[derive(Clone, Debug)]
+struct ModuleMetadata {
+  specifier: ModuleSpecifier,
+  media_type: MediaType,
+}
+
 struct Builder<'a, 'graph> {
   in_dynamic_branch: bool,
   graph: &'graph mut ModuleGraph,
   loader: &'a mut dyn Loader,
   resolver: Option<&'a dyn Resolver>,
   pending: FuturesUnordered<LoadWithSpecifierFuture>,
-  pending_specifiers:
-    HashMap<ModuleSpecifier, HashSet<Option<AssertTypeWithRange>>>,
-  dynamic_branches:
-    HashMap<ModuleSpecifier, (Range, Option<AssertTypeWithRange>)>,
+  dynamic_branches: HashMap<ModuleSpecifier, Range>,
   module_analyzer: &'a dyn ModuleAnalyzer,
   reporter: Option<&'a dyn Reporter>,
+  module_metadata: HashMap<ModuleSpecifier, ModuleMetadata>,
```

Comment:
I'm not sure we need this because this information should be avaiable straight from the graph? I think this will maybe fail if someone calls `build` multiple times.

Reply:
It's to work around borrow issues at https://github.com/denoland/deno_graph/blob/af1009eb0e5cd9dcb1c851911f48268ed187f7b5/src/graph.rs#L1877-L1885
If there were a workaround limited to `Builder::fill()`, it would probably be to build up the same map but as its own pass just above this.

Reply:
Yeah, we'll have to make that work based on the current module graph. It would be better if we didn't create two sources of truth and just derived from the graph.